### PR TITLE
Implement challenge detail page design

### LINKED ISSUE #117: Create wireframe for the Challenge page
A placeholder component will be provided by #108

### PR DESCRIPTION
**Pull request by rrchai**
FIxes #117: Implement the wireframe design on challenge page from [Figma wireframe](https://www.figma.com/file/UEFKtKBe35pcPeZHcx0sKL/Challenge-Profile):

- [x] Add layout: header (banner + title + dates ...) + content (overview + organizers + organizations)
- [x] Implement styles based on the wireframe design
- [x] Add material progress bar
- [x] Add favorite button
- [x] Add media styles

recording

### CODE REVIEW

**Comment by tschaffter** [comment]
@rrchai Try `*ngIf="challenge$ | async"`. Observables like `challenge$` need the async pipe when used in the HTML document.

**Comment by rrchai** [comment]
@vpchung Once you have your new design, we could start a new PR and modify based on this draft.

File: src/app/pages/challenges/challenge-view/challenge-view.component.ts [hunk 1]
```diff
@@ -14,22 +20,77 @@ import { PageTitleService } from '@shared/page-title/page-title.service';
...
-      switchMap((params) => this.challengeService.getChallenge(params.id))
+    // this.challenge$ = this.route.params.pipe(
+    //   switchMap((params) => this.challengeService.getChallenge(params.id))
+    // );
+    // this.challenge$.subscribe(challenge => {
+    //   this.pageTitleService.setTitle(challenge.name);
+    // });
+    this.route.params.pipe(
+      switchMap((params) => this.challengeService.getChallenge(params.id)),
+      tap(challengeRes => {
```
Comment (tschaffter) [inline src/app/pages/challenges/challenge-view/challenge-view.component.ts:52]:
```suggestion
      tap(challenge => {
```

File: src/app/pages/challenges/challenge-view/challenge-view.component.ts [hunk 2]
```diff
@@ -14,22 +20,77 @@ import { PageTitleService } from '@shared/page-title/page-title.service';
...
+        this.challenge = challengeRes;
+        this.pageTitleService.setTitle(challengeRes.name);
+      }),
+      switchMap(challenge => 
+        forkJoin(
+          challenge.organizerIds
+            .map(organizerId => this.personService.getPerson(organizerId))
+        )
+      ),
+      tap((persons)=> this.personList = persons),
```
Comment (tschaffter) [inline src/app/pages/challenges/challenge-view/challenge-view.component.ts:62]:
```suggestion
      tap((persons) => this.personList = persons),
```

Comment (tschaffter) [reply src/app/pages/challenges/challenge-view/challenge-view.component.ts:62]: Run the document formatter to fix this and potentially other style typos.

File: src/app/pages/challenges/challenge-view/challenge-view.component.ts [hunk 3]
```diff
@@ -14,22 +20,77 @@ import { PageTitleService } from '@shared/page-title/page-title.service';
...
+      switchMap((persons) => {
+        const orgIds = [...new Set(persons.flatMap(person => person.organizationIds))]
```
Comment (tschaffter) [inline src/app/pages/challenges/challenge-view/challenge-view.component.ts:64]: Why not use `Challenge.organizations`? The issue with the current implementation is that 10 years from now, the organizers will be working for other organization and the name of these orgs will be displayed on the page of old challenges, while these organizations have in no way contributed to the organization of the challenge.

File: tsconfig.app.json [hunk 4]
```diff
@@ -3,7 +3,10 @@
   "extends": "./tsconfig.json",
   "compilerOptions": {
     "outDir": "./out-tsc/app",
-    "types": []
+    "types": [],
+    "target": "es2018",
+    "module": "es2020",
+    "lib": ["es2020", "dom"],
```
Comment (tschaffter) [inline tsconfig.app.json:9]: Why is this required?

**Review by tschaffter — CHANGES_REQUESTED** [review]
Additional comments:

- Ideally align horizontally to the left the left side of a section title and the left side of its content. Unless this is part of the Card initial design.

image

- Consider using a smaller font size for the challenge description. We should not use more than ~3 font sizes across the entire app.

**Comment by vpchung** [comment]
This is awesome!!  Nicely done, @rrchai !

Upon seeing the wireframe "in action", there are a couple things I'd like to see different, including:

- more padding under the timeline progress bar
- something to make the challenge status more obvious
- moving the tags to a different location?
- grid display for the list of organizers
- etc.

I'll start designing later today!

**Review by vpchung — COMMENTED** [review]
This is awesome!!! Nicely done, @rrchai !

Upon seeing the wireframe "in action", there are a couple things I'd like to see different, including:

- more padding under the timeline progress bar
- something to make the challenge status more obvious
- move the tags to a different location?
- grid display for the list of organizers
- etc.

I'll start designing later today!

File: src/app/pages/challenges/challenge-view/challenge-view.component.ts (hunk 3 above)
Comment (tschaffter) [reply src/app/pages/challenges/challenge-view/challenge-view.component.ts:64]: I just noticed that `Challenge.organizations` is not (yet) a thing. This property will be added in an update of the schemas coming this week. For now let's not show organizations on the challenge page for simplicity.

File: src/app/pages/challenges/challenge-view/challenge-view.component.ts (hunk 1 above)
Comment (rrchai) [reply src/app/pages/challenges/challenge-view/challenge-view.component.ts:52]: >Ideally align horizontally to the left the left side of a section title and the left side of its content. Unless this is part of the Card initial design.

It was card's original design, but I will change to make them align.

File: tsconfig.app.json (hunk 4 above)
Comment (rrchai) [reply tsconfig.app.json:9]: The change was to prevent from raising errors below for `flatMap`:
```
Property 'flatMap' does not exist on type 'Person[]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2019' or later
```

Since we will get organizations from challenge object soon, I will just remove organization part (where I used `flatMap`) and change `tsconfig.app.json` this back to what it was.
